File: examples/ts/token-ft/FungibleTokenExample.tsop.ts
import { StatefulSmartContract, assert, checkSig } from 'tsop-lang';
import type { PubKey, Sig, ByteString } from 'tsop-lang';

class FungibleToken extends StatefulSmartContract {
  owner: PubKey;           // stateful: current token owner
  balance: bigint;         // stateful: token balance in this UTXO
  readonly tokenId: ByteString; // immutable: token identifier

  constructor(owner: PubKey, balance: bigint, tokenId: ByteString) {
    super(owner, balance, tokenId);
    this.owner = owner;
    this.balance = balance;
    this.tokenId = tokenId;
  }

  // Split: 1 input → 2 outputs (recipient + change)
  public transfer(sig: Sig, to: PubKey, amount: bigint, outputSatoshis: bigint) {
    assert(checkSig(sig, this.owner));
    assert(amount > 0n);
    assert(amount <= this.balance);

    // addOutput(satoshis, owner, balance) — args match mutable props in order
    this.addOutput(outputSatoshis, to, amount);
    this.addOutput(outputSatoshis, this.owner, this.balance - amount);
  }

  // Simple send: 1 input → 1 output, full balance
  public send(sig: Sig, to: PubKey, outputSatoshis: bigint) {
    assert(checkSig(sig, this.owner));

    this.addOutput(outputSatoshis, to, this.balance);
  }

  // Merge: N inputs → 1 output (each input calls this independently)
  public merge(sig: Sig, totalBalance: bigint, outputSatoshis: bigint) {
    assert(checkSig(sig, this.owner));
    assert(totalBalance >= this.balance);

    this.addOutput(outputSatoshis, this.owner, totalBalance);
  }
}

File: examples/go/token-ft/FungibleTokenExample.tsop.go
package contract

import "tsop"

type FungibleToken struct {
	tsop.StatefulSmartContract
	Owner   tsop.PubKey     // stateful: current token owner
	Balance tsop.Bigint     // stateful: token balance in this UTXO
	TokenId tsop.ByteString `tsop:"readonly"` // immutable: token identifier
}

// Split: 1 input -> 2 outputs (recipient + change)
func (c *FungibleToken) Transfer(sig tsop.Sig, to tsop.PubKey, amount tsop.Bigint, outputSatoshis tsop.Bigint) {
	tsop.Assert(tsop.CheckSig(sig, c.Owner))
	tsop.Assert(amount > 0)
	tsop.Assert(amount <= c.Balance)

	// AddOutput(satoshis, owner, balance) -- args match mutable props in order
	c.AddOutput(outputSatoshis, to, amount)
	c.AddOutput(outputSatoshis, c.Owner, c.Balance-amount)
}

// Simple send: 1 input -> 1 output, full balance
func (c *FungibleToken) Send(sig tsop.Sig, to tsop.PubKey, outputSatoshis tsop.Bigint) {
	tsop.Assert(tsop.CheckSig(sig, c.Owner))

	c.AddOutput(outputSatoshis, to, c.Balance)
}

// Merge: N inputs -> 1 output (each input calls this independently)
func (c *FungibleToken) Merge(sig tsop.Sig, totalBalance tsop.Bigint, outputSatoshis tsop.Bigint) {
	tsop.Assert(tsop.CheckSig(sig, c.Owner))
	tsop.Assert(totalBalance >= c.Balance)

	c.AddOutput(outputSatoshis, c.Owner, totalBalance)
}
